katcp/object_proxies.py

```diff
@@ -0,0 +1,187 @@
+            "    return ob %(op)s self.__subject__\n"
+            "\n"
+            "def __i%(name)s__(self,ob):\n"
+            "    self.__subject__ %(op)s=ob\n"
+            "    return self\n"
+        )  % locals())
+
+    del name, op
+
+    # Oddball signatures
+
+    def __rdivmod__(self,ob):
+        return divmod(ob, self.__subject__)
+
+    def __pow__(self,*args):
+        return pow(self.__subject__,*args)
+
+    def __ipow__(self,ob):
+        self.__subject__ **= ob
+        return self
+
+    def __rpow__(self,ob):
+        return pow(ob, self.__subject__)
+
+
+class ObjectProxy(AbstractProxy):
+    """Proxy for a specific object"""
+
+    __slots__ = "__subject__"
+
+    def __init__(self,subject):
+        self.__subject__ = subject
+
+
+class CallbackProxy(AbstractProxy):
+    """Proxy for a dynamically-chosen object"""
+
+    __slots__ = '__callback__'
+
+    def __init__(self, func):
+        set_callback(self,func)
+
+set_callback = CallbackProxy.__callback__.__set__
+get_callback = CallbackProxy.__callback__.__get__
+CallbackProxy.__subject__ = property(lambda self, gc=get_callback: gc(self)())
+
+
+
+class LazyProxy(CallbackProxy):
+    """Proxy for a lazily-obtained object, that is cached on first use"""
+    __slots__ = "__cache__"
+
+get_cache = LazyProxy.__cache__.__get__
+set_cache = LazyProxy.__cache__.__set__
+
+def __subject__(self, get_cache=get_cache, set_cache=set_cache):
+    try:
+        return get_cache(self)
+    except AttributeError:
+        set_cache(self, get_callback(self)())
+        return get_cache(self)
+
+LazyProxy.__subject__ = property(__subject__, set_cache)
+del __subject__
+
+
+class AbstractWrapper(AbstractProxy):
+    """Mixin to allow extra behaviors and attributes on proxy instance"""
+    __slots__ = ()
+
+    def __getattribute__(self, attr, oga=object.__getattribute__):
+        if attr.startswith('__'):
+            subject = oga(self,'__subject__')
+            if attr=='__subject__':
+                return subject
+            return getattr(subject,attr)
+        return oga(self,attr)
+
+    def __getattr__(self,attr, oga=object.__getattribute__):
+        return getattr(oga(self,'__subject__'), attr)
+
+    def __setattr__(self,attr,val, osa=object.__setattr__):
+        if (
+            attr=='__subject__'
+            or hasattr(type(self),attr) and not attr.startswith('__')
+        ):
+            osa(self,attr,val)
+        else:
+            setattr(self.__subject__,attr,val)
+
+    def __delattr__(self,attr, oda=object.__delattr__):
+        if (
+            attr=='__subject__'
+            or hasattr(type(self),attr) and not attr.startswith('__')
+        ):
+            oda(self,attr)
+        else:
+            delattr(self.__subject__,attr)
+
+class ObjectWrapper(ObjectProxy, AbstractWrapper):      __slots__ = ()
+class CallbackWrapper(CallbackProxy, AbstractWrapper):  __slots__ = ()
+class LazyWrapper(LazyProxy, AbstractWrapper):          __slots__ = ()
```

Comment:
I thought about that too, but since it is basically a copy of an upstream module it seemed to make sense to leave it as close to original as possible. If, however, someone went to the effort of making a PEP8 PR for that I would not argue :)